p2p/dandelion/src/txpool.rs

```diff
@@ -0,0 +1,545 @@
+//! # Dandelion++ TxPool
+//!
+//! This is an implementation of a dandelion++ compatible tx pool.
+//!
+use std::{
+    collections::{HashMap, HashSet},
+    future::Future,
+    hash::Hash,
+    marker::PhantomData,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use futures::{FutureExt, StreamExt};
+use rand::prelude::*;
+use rand_distr::Exp;
+use tokio::{
+    sync::{mpsc, oneshot},
+    task::JoinSet,
+};
+use tokio_util::{sync::PollSender, time::DelayQueue};
+use tower::{Service, ServiceExt};
+use tracing::Instrument;
+
+use crate::{
+    traits::{TxStoreRequest, TxStoreResponse},
+    DandelionConfig, DandelionRouteReq, DandelionRouterError, State, TxState,
+};
+
+#[derive(Copy, Clone, Debug, thiserror::Error)]
+#[error("The dandelion pool was shutdown")]
+pub struct DandelionPoolShutDown;
+
+/// A request to the dandelion tx pool.
+///
+/// This enum does not contain requests to get tx data from the pool, this is to reduce the amount
+/// of requests that get sent to the [`DandelionPool`]. Users should hold a handle to the backing tx storage
+/// to get transaction data from the pool.
+///
+/// It MUST be remembered to not expose the private stem pool to peers.
```

Comment:
Seems like the type system should be used here...

Reply:
We can't enforce this on outside crates without requiring all requests to the txpool to go through the dandelion pool, which would add unnecessary indirection.

Even then crates could still misuses this this.

Cuprate will enforce it in it's own tx-pool but it's not something we can do here